src/python/pants/core/goals/publish.py

```diff
@@ -0,0 +1,160 @@
+from dataclasses import dataclass
+import logging
+from typing import ClassVar, Iterable, List, Tuple, Type, cast
+
+from pants.build_graph.address import Address
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.engine.addresses import UnparsedAddressInputs
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import MultiGet
+from pants.engine.rules import Get, collect_rules, goal_rule
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    FieldSet,
+    SpecialCasedDependencies,
+    Target,
+    Targets,
+)
+from pants.engine.unions import UnionMembership, union
+from pants.util.meta import frozen_after_init
+from pants.util.ordered_set import FrozenOrderedSet
+
+logger = logging.getLogger(__name__)
+
+
+class PublishTargetField(SpecialCasedDependencies):
+    alias = "publish_targets"
+    help = "Repositories to publish to."
+
+
+@dataclass(frozen=True)
+class PublishTargetFieldSet(FieldSet):
+    required_fields = (PublishTargetField,)
+
+
+@union
+@dataclass(frozen=True)
+class PublishRequest:
+    built_package: BuiltPackage
+    fieldset: PublishTargetFieldSet
+    publish_target: "PublishTarget"
+
+
+class PublishTarget(Target):
+    core_fields = (*COMMON_TARGET_FIELDS,)
+
+    publish_request_type: ClassVar[Type[PublishRequest]]
+    publishee_fieldset_type: ClassVar[Type[PublishTargetFieldSet]] = PublishTargetFieldSet
+
+
+class PublishSubsystem(GoalSubsystem):
+    name = "publish"
+    help = "Publish packages."
+
+
+class Publish(Goal):
```

Comment:
Do I need to register this anywhere to make `./pants publish` work?

Reply:
Yes, I think you need to call the `rules()` method in here from the `rules()` method over there: https://github.com/pantsbuild/pants/blob/main/src/python/pants/core/register.py#L27